### PR TITLE
Improvement of sync_on_save

### PR DESCRIPTION
**Pull request by vijayasantham12**
Hi @fyears, I just improved sync_on_save to avoid two minor problems

- If I continuously write without break, it won't sync the last word(sometimes few characters). Note this issue occur on sync_on_save every 1 sec. In this PR, it will sync all the content even we type super fast without break
- I think check every `syncOnSaveAfterMilliseconds` secs to sync the changes is similar to "Schedule for auto run" so I removed the interval and add event listener to listen file changes

But still, we should set 1/5/10 seconds on the sync_on_save settings to schedule sync(it's not like check every 1/5/10 secs, just setTimeout) and avoid making too many API requests to the sync provider.

### CODE REVIEW

**Comment by fyears** [comment]
Hi thanks for your PR.

... but could you please explain the solution in the PR? I am feeling confused, are you trying to delay the sync if the file is changed within 1 sec?

**Comment by vijayasantham12** [comment]
> are you trying to delay the sync if the file is changed within 1 sec?

No, I am not attempting to delay the sync if the file is changed within 1 sec. Instead, the `checkCurrFileModified` function is called whenever a file change is saved to disk (vault-modify event), and it is called again once the sync is completed. Basically we are running `checkCurrFileModified` twice for a single file change.

Now, let me explain why:

Let's consider a scenario where User A writes "A" in a note. This triggers a vault-modify event, causing `checkCurrFileModified` to run and schedule `scheduleSyncOnSave` after 1 second (or based on the user's settings). 

The sync process typically takes around 1-2 secs (depending on vault size and network speed). 

If User A then types "AB" in the note while the sync is in progress, another vault-modify event will trigger, causing `checkCurrFileModified` to run again. However, this time, it will not schedule a sync. Instead, it will set `needToRunAgain` to true because we schedule a sync if either `syncStatus === "idle" & !runScheduled & lastModified > lastSuccessSyncTime` or `syncStatus === "idle" & !runScheduled & needToRunAgain`.

So, when the sync is completed, the `SYNC_DONE` event will trigger. Now, the condition `syncStatus === "idle" & !runScheduled & needToRunAgain` will pass because `needToRunAgain` is `true`.

In summary, the process ensures that if a file is changed during the sync process, another sync is scheduled to capture those changes.

**Comment by kadisonm** [comment]
Hi @fyears will this be added? I think 'this.app.vault.on("modify")' is less expensive than checking at an interval. 
Correct me if I'm wrong though.

**Comment by fyears** [comment]
Admittedly the original implementation is sub-optimal and over-complicated. I am think of a better way. Actually I don't   understand the usage of syncOnSaveAfterMilliseconds, which I think it's unnecessary if we use on vault modify
